### DISCUSSION TITLE
NRules suitability for a new project

### DISCUSSION THREAD

**Discussion by yaronbsmtx** [NRules/NRules#359] [General]
Hi,

We are considering using NRules for a new project.
The core function of this project is to find which items - from a list of a few hundred - can be said to be appropriate for a user depending on answers he/she provides for a dynamic set of questions.
Each item has a set of rules - based on mathematical operators (>, < and so on) and value lists, possibly grouped and having logical operators between them (and, or) - such that if they are met, the item is found to be appropriate.
Each set can have a varying number of rules, possibly up to ~100 in rare cases.

For example, the system may have this rule set for the item "get a puppy":

1. If you do not have a puppy yet
AND
2. If you are between the ages 3 and 18

and this rule set for the item "get a car":
1. If you have a driver license
AND
(2a. If you are between the ages 18 and 65
OR
2b. You are Batman)

When a user answers the question "Do you have a puppy" with a "No", no item is found to be appropriate.
If, for the next question "Do you have a driver license", the answer is "Yes", still no item is found to be appropriate.
If, for the next question "How old are you", the answer is "15", the first item is the only one found to be appropriate.
If, for the last question "Are you Batman", the answer is "Yes", both items are found to be appropriate.

NRules seems to be a possible solution for this type of challenge, but I would like to know - based on the community's experience - how well is it suited for such a project and what obstacles should we look out for with regards to implementation and performance.

**Comment by snikolayev** [comment]
@yaronbsmtx I'm obviously biased, but I think NRules is quite suited for expressing the logic you are describing. I really suggest to do a small POC and get a feel for it.

The number of facts you are quoting are pretty small, so I wouldn't expect you to worry about performance.

What you would need to think about is the overall application architecture and how NRules fits into it. E.g. where rules will live vs runtime of the engine that discovers and compiles the rules, converts user input into facts and feeds those facts to the engine. This may look very different, depending on the type of the application (e.g. a simple console app vs a backend service vs a web app, etc.). Also, rules are written in terms of the domain model, so you need to model your domain well, such that it's easy and natural to express your desired logic.
